### DISCUSSION TITLE
Multi-select Combobox

### DISCUSSION THREAD

**Discussion by pguptaDS** [twilio-labs/paste#1261] [General] [answered]
Hi! RBAC team is exploring a solution where a user can select multiple permissions from a list of permissions. We will be building this functionality in Q2 so I'd love to explore how we can build this functionality using components that are available now. Here's an image that communicates the functionality we'd like to have:

![image](https://user-images.githubusercontent.com/64559618/110179766-7e870180-7dbd-11eb-8895-418cb21af416.png)

**Comment by nkrantz** [comment, marked as answer]
Hi @pguptaDS, thanks for asking about multi-select combobox. There's currently no way to build out your desired functionality on top of our combobox (component nor primitive). The multi-select combobox component is on our [roadmap](https://paste.twilio.design/roadmap/), scheduled to be released in July.

For now, check out [Katie's answer to a similar question](https://github.com/twilio-labs/paste/discussions/939). The best option for now is probably to use a [checkbox group](https://paste.twilio.design/components/checkbox/#unchecked-checked-and-indeterminate). Would this work for your use-case until the component is released, or would there be too many options/permissions to select from?

If the multi-select combobox is the only path forward, we can talk about next steps for getting it ready to use before July. We’d have to add the [useMultipleSelection hook](https://downshift.netlify.app/use-multiple-selection) to our combobox primitive so that your team could build out the component. We’d work alongside you to ensure that it’s built in an accessible and sustainable way. If this is something you want to consider, please let me know so I can share with the team.

**Reply by pguptaDS** [reply]
Hi @nkrantz! There will definitely be a long list of permissions to select from. I think we will need some sort of a container that allows search functionality as well since manually scrolling to find the permission you're looking for in a long list is not  user-friendly. That is why we were exploring a combobox. Is it possible to wrap the checkbox group in a combobox? If not, we might have to discuss using the hook that you mentioned. What would be a good way to have that conversation?

**Reply by nkrantz** [reply]
@pguptaDS thanks for the clarification! Before we talk about next steps, I want to get a better sense of what you're trying to achieve here. We may have a good path forward using our current components. What's the most important aspect of your desired functionality? Is it to show users what they've already selected? The ability to search through the list by typing in the input field? Some more context here about what exactly you're aiming for will help us think through an appropriate solution.

**Reply by pguptaDS** [reply]
@nkrantz Sure. Here's the functionality we need to incorporate: 

1. View list of all available permissions on the platform. We would like to show permission name & their description. A permission name looks like this /twilio/studio/flows/create. Every permission has this format wherein it has 4 components separated by a /
2. Have the ability to select multiple permissions for their API Key
3. Have the ability to search for permissions by typing keywords such as studio or flows
4. Have the ability to bulk select permissions & see the selection by using wildcard support.. for example a user could type /twilio/studio/* and we would select all studio permissions in the box shown. Similarly a user can further type /twilio/voice/** and we would select all the voice permissions. 

For context, the reason why wildcard support is a must have is because our platform will have a large number of permissions.
 - There are use cases wherein customers want a lot of granularity to control what permission would be allowed, for example, only read & create sms permissions but not delete sms… in this case they would just search & select.
- There are also use cases wherein a customers wants broader access, likely by products… so they would limit access of their API key using the wildcard support.

Hope that helps.

**Reply by nkrantz** [reply]
That's super helpful context @pguptaDS, thanks for sharing. After talking with the team, we have some ideas of how to proceed, but the wildcard search aspect does increase the complexity significantly. Have you tested these designs with users or do you plan to?

**Reply by pguptaDS** [reply]
@nkrantz  Glad to hear that. We have not tested anything yet as I wanted to bring in the Paste team in the process sooner than later. We are actually planning on running customer interviews in the week of Mar 22 and wanted to show some possible concepts of the functionality for validation during that research. I tried to sign up for DSys office hours but it seems like the first available slot is April 8 which is too late for us (didn't realize there's a month long waitlist for OH now!). Perhaps we can schedule a one-off session so I can get started on creating some designs? I am not sure what the best route here is.

**Reply by serifluous** [reply]
Hey @pguptaDS !

We put together a CodeSandbox example that might help you directionally and doesn't use a combobox at all: https://codesandbox.io/s/cold-sound-9ylmj?file=/src/index.tsx It's not fully functional, but the idea is that you could put checkbox groups in a scrollable card, and have a text input above it that filters through the groups.

Specifically with the wildcard search and bulk selection requirements, even the multiselect combobox we build near-term might not be able to support that. Our first release will likely be closer to this, an extension of our current combobox but allows for and shows multiple selections: https://www.figma.com/file/y1cYRhZEApY1J9HvUT3NzZ/react-components-migration?node-id=5%3A5434

If you end up testing out wildcard search, we'd love to learn about your findings if you can reshare here. It's not functionality we've seen across the products before, but if it's a pressing customer need and straightforward for them to discover/use, we definitely want to know and consider it for future improvements.

Some other similar use cases we have seen, in case they help jog further ideas:

- Discussions #1153 and #1271 (in progress) from @mindytom where a customer needs to find and select lots of different events by product
- Roles & permissions in Conversations from Kevin H https://share.goabstract.com/20c4a326-a22f-406d-b12e-b32777ac005e

**Reply by pguptaDS** [reply]
@serifluous Thank you so much for this! This was super helpful to visualize. Looking at the multi-select functionality that you linked here, my PM and I almost prefer the sandbox layout more for our use case. This gives me enough to go forward and put together designs and prototype ahead of our research. We are planning on making part of our customer interview into a usability testing session as well so if we get a chance to talk about wildcard search, I'd be sure to bring it back to this group. We are also meeting up with our FE lead to figure out the backend capabilities to support wildcard search. Stay tuned on all of that.

Thank you again for the prompt guidance here.

**Reply by serifluous** [reply]
Nice, we're glad to hear! And props to @richbachman for putting together that CodeSandbox so quickly.

**Reply by kendal-kirby** [reply]
Hi @nkrantz, I'm a new designer on the team also in need of a multi-select combobox, so I wanted to follow up on this thread. Is there still a plan to release this in July? I'm not seeing it on the current roadmap. Thanks!

**Reply by richbachman** [reply]
Hi @kendal-kirby. Welcome and thanks for posting! 

Currently, the multi-select combobox is slated for November 16 release (Release - 2021.11.16 TBD): https://paste.twilio.design/roadmap/
